Implement basic rerecording features

zhaowenlan1779 opened the pull request:
Closes #5387.

This implements some rerecording features (basically, rerecording is about how savestates interact with movies). When developing these features, I used TASVideos's [requirements](http://tasvideos.org/EmulatorResources/Requirements.html) and [desired features](http://tasvideos.org/EmulatorResources/Features.html) for reference, and also got advice from some of their staff.

Brief description of what this does (more comments in commit message):
1. Fixed critical desync and file corruption bugs
1. You now have a separate set of savestate slots for each movie you are doing. This is provided for convenience and also to prevent loading states for other movies
1. Added a read-only mode. When read-only mode is on, loading a state resumes playback from that point; when read-only mode is off, loading a state **saves the current movie being recorded, restores the movie data stored in the savestate** and starts recording from that point (rerecording)
1. Added author, rerecord count and input count (aka length)
1. Remade movie UI

What I also want but will probably not happen very soon:
1. Somehow save the HLE applet inputs (e.g. the text you input in swkbd)

---

Screenshots:
![image](https://user-images.githubusercontent.com/21307832/86770384-33c4b680-c083-11ea-9a12-fd4ab1802c6d.png)
![image](https://user-images.githubusercontent.com/21307832/86770421-3d4e1e80-c083-11ea-9fde-ec55629c248f.png)
![image](https://user-images.githubusercontent.com/21307832/86770492-5c4cb080-c083-11ea-9ff3-c54197367f73.png)

InfamousKnight commented:
Nice to see this! I did some testing with driver renegade, it syncs just fine.

Now a couple of minor problems I have with this: input is not shown on the status bar. So if I'm holding frame advance while holding the B button, it would say B on that frame. Plus any other inputs being pressed at that time. Like left or right.

The second minor issue I have, is when youre holding frame advance and holding B and like left, and then you let go of the left key and press right while still holding B and frame advance, it would pause the emulator until I let go of frame advance and press it again to continue with the right key. It would be nice if it didn't pause.

phoenix1291 commented:
Requests list from Citra topic on Tasvideos

- option to start emulation (game) paused
- option to pause emulation when movie is finished
- option and hotkeys to save/load current savestate (not save to oldest slot and load from newest slot)
- hotkeys to backup/load on the same slot and choose the backup slot (like 1,2,3,4,5,6,7,8,9,o)
- option to save named state (exported savestate, like in Bizhawk, not in the 10 slots)
- osd display of wich slot of save is in use
- option to save the movie at any time (like in Bizhawk again)
- implementation of Lagarith and/or Camstudio lossless codecs
- option to dump the video and the audio in separate files
- osd display the buttons that were pressed when playing and while playing back a movie
- osd display for when a savestate is saved/loaded
- rewind feature

To fix:
- 3DS internal keyboard doesn't always work with LLE applets with games imported into Citra from the 3DS with treeSD (for example RE Revelations and Mario & Sonic London 2012 always offers me the Windows prompt to enter the name and not the 3DS keyboard)


More to come?

FearlessTobi commented:
Needs a rebase, sorry.

zhaowenlan1779 commented:
> option to start emulation (game) paused
> option to pause emulation when movie is finished

Implemented.

> option and hotkeys to save/load current savestate (not save to oldest slot and load from newest slot)
> hotkeys to backup/load on the same slot and choose the backup slot (like 1,2,3,4,5,6,7,8,9,o)
> option to save named state (exported savestate, like in Bizhawk, not in the 10 slots)

These will probably be in a different PR

> osd display of wich slot of save is in use
> osd display the buttons that were pressed when playing and while playing back a movie
> osd display for when a savestate is saved/loaded

Eh, since Citra doesn't have an overlay yet these will be quite some work and probably won't happen soon.

> option to save the movie at any time (like in Bizhawk again)

Implemented.

> implementation of Lagarith and/or Camstudio lossless codecs
> option to dump the video and the audio in separate files

Not sure why these would be useful?

> rewind feature

The 3DS is too modern for this to be feasible

vadosnaprimer commented:
Ability to dump lossless video is essential to TASing:
http://tasvideos.org/LawsOfTAS/OnVideoDumping.html

Another good codec option is ffv1.

zhaowenlan1779 commented:
Citra already supports a wide range of codecs, including ffv1.

InfamousKnight commented:
About rewind, all it does is constantly make save states every frame(or every x frame for performance), and when rewinding it just loads save states on a counter. It shouldn't be too hard once save states even exist.

B3n30 commented:
@InfamousKnight Just fyi save states already exists, but for modern systems you have to store that much of data that it slows down the emulation so you can't continously create save states and continue playing without frame drops

phoenix1291 commented:
Urgent:
- Microphone: hotkey for this like in DeSmuME
- Savestate: hotkey to save an load current slot (in addition to being able to save savestates by name outside of the 10 basic slots provided (as in Bizhawk))
- Screen position: a native solution for "larger sceen" to choose % of the botoom screen? I think of 2x the actual size, but having more solutions is good too. And can choose if the "botton screen" is bottom, middle or top, natively... The actual size is way too small.
- Virtual pad: For games like Kid Icarus, it would be great to have a "Virtual Pad", like in Bizhawk
- Movie: How to recover the content of the movie to be able to paste it in a new one or change which version to launch in the movie file (header?)?

Thanks!

B3n30 commented:
LGTM

zhaowenlan1779 commented:
This PR still has some issues I intend to investigate, but I probably won't have the chance to until mid June. Sorry about that.

InfamousKnight commented:
> @InfamousKnight Just fyi save states already exists, but for modern systems you have to store that much of data that it slows down the emulation so you can't continously create save states and continue playing without frame drops

If its being used for tasing, it running slow would be beneficial. Obliviously too slow like 1fps wouldnt be fun.

InfamousKnight commented:
Nitpick: under frame advancing: if you press a different button while holding frame advance, it will pause until you let go of frame advance and press it again with the new set of keys.

So for instance: holding down A while frame advance is enabled, then you hold down left or right, it will pause the game until you let go of frame advance hot key, and press it again. I think you have to let go of all keys and press them again. It just feels weird.

phoenix1291 commented:
What changes have been made since July 2021?
I haven't seen any changes in the UI or in the functions implemented in the latest Canary versions.
I would like to do some more testing, but it's hard to know what has been improved, implemented or not.
